kstatus/wait/wait.go

```diff
@@ -10,28 +10,59 @@ import (
 
 	"github.com/pkg/errors"
 	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/kustomize/kstatus/status"
 )
 
+const (
+	defaultNamespace = "default"
+)
+
 // ResourceIdentifier defines the functions needed to identify
 // a resource in a cluster. This interface is implemented by
 // both unstructured.Unstructured and the standard Kubernetes types.
-type ResourceIdentifier interface {
+type KubernetesObject interface {
 	GetName() string
 	GetNamespace() string
-	GetAPIVersion() string
-	GetKind() string
+	GroupVersionKind() schema.GroupVersionKind
+}
+
+// ResourceIdentifier contains the information needed to uniquely
+// identify a resource in a cluster.
+type ResourceIdentifier struct {
+	Name      string
+	Namespace string
+	GroupKind schema.GroupKind
+}
+
+// Equals compares two ResourceIdentifiers and returns true if they
+// refer to the same resource. Special handling is needed for namespace
+// since an empty namespace for a namespace-scoped resource is defaulted
+// to the "default" namespace.
+func (r ResourceIdentifier) Equals(other ResourceIdentifier) bool {
+	isSameNamespace := r.Namespace == other.Namespace ||
+		(r.Namespace == "" && other.Namespace == defaultNamespace) ||
+		(r.Namespace == defaultNamespace && other.Namespace == "")
```

Comment:
the notion of "resource ID" and equality tests for it, Namespace, GK, GVK, etc. depend on the context, and if they were all in one place, we'd have an easier time managing them.

Would you mind if this moved to
 https://github.com/kubernetes-sigs/kustomize/blob/master/api/resid/resid.go later?

Reply:
I agree on this. I think we can combine this with resid.

Reply:
OK, doesn't have to happen right now